lib/queuePopulator/QueuePopulator.js

```diff
@@ -11,6 +9,102 @@ const MetricsProducer = require('../MetricsProducer');
+const promClient = require('prom-client');
+const metricLabels = ['origin', 'logName', 'logId', 'containerName'];
+promClient.register.setDefaultLabels({
+    origin: 'replication',
+    containerName: process.env.CONTAINER_NAME || '',
+});
+
+/**
+ * Labels used for Prometheus metrics
+ * @typedef {Object} MetricLabels
+ * @property {string} logName - Name of the log we are using
+ * @property {string} logId - Id of the log we are reading
+ * @property {string} containerName - Name of the container running our process
+ * @property {string} [publishStatus] - Result of the publishing to kafka to the topic
+ */
+
+/**
+ * Increment a Prometheus counter metric
+ * @typedef { (labels: MetricLabels, value?: number) => void } CounterInc
+ */
+
+/**
+ * Set a Prometheus gauge metric
+ * @typedef { (labels: MetricLabels, value: number) => void } GaugeSet
+ */
+
+/**
+ * WrapCounterInc wraps the Counters Inc method adding the metric labels used.
+ *
+ * @param {promClient.Counter} metric - Prom counter metric
+ * @returns {CounterInc} - Function used to increment counter
+ */
+function wrapCounterInc(metric) {
+    return (labels, value) => {
+        metric.inc(labels, value);
+    };
+}
+
+/**
+ * Wrap Gauge Set wraps a Prometheus Guage's set method
+ *
+ * @param {promClient.Gauge} metric - Prom gauge metric
+ * @returns {GaugeSet} - Function used to set gauge
+ */
+function wrapGaugeSet(metric) {
+    return (labels, value) => {
+        metric.set(labels, value);
+    };
+}
+
+const logReadOffsetMetric = new promClient.Gauge({
+    name: 'replication_read_offset',
+    help: 'Current read offset of metadata journal',
+    labelNames: metricLabels,
+});
+
+const logSizeMetric = new promClient.Gauge({
+    name: 'replication_log_size',
+    help: 'Current size of metadata journal',
+    labelNames: metricLabels,
+});
+
+const messageMetrics = new promClient.Counter({
+    name: 'replication_populator_messages',
+    help: 'Total number of Kafka messages produced by the queue populator',
+    labelNames: ['origin', 'logName', 'logId', 'containerName', 'publishStatus'],
+});
+
+const objectMetrics = new promClient.Counter({
+    name: 'replication_populator_objects',
+    help: 'Total objects queued for replication',
+    labelNames: metricLabels,
+});
+
+const byteMetrics = new promClient.Counter({
+    name: 'replication_populator_bytes',
+    help: 'Total number of bytes queued for replication not including metadata',
+    labelNames: metricLabels,
+});
+
+/**
+ * Contains methods to incrememt different metrics
+ * @typedef {Object} MetricsHandler
+ * @property {CounterInc} messages - Increments the message metric
+ * @property {CounterInc} objects - Increments the objects metric
+ * @property {CounterInc} bytes - Increments the bytes metric
+ * @property {GaugeSet} logReadOffset - Set the log read offset metric
+ * @property {GaugeSet} logSize - Set the log size metric
+ */
+const metricsHandler = {
```

Comment:
Should metricsHandler be a class?

Reply:
Metrics handler is just a collection of metric methods, there is no internal state or parameters to pass so I feel a class is unnecessary.

Reply:
are the functions reusable for processor, we dont want to duplicate!

Reply:
the GaugeSet and CounterInc methods and types will be for sure, maybe a metrics util file will come in handy. metricsHandler itself will need to be per process but it should be relatively similar across the board.

Reply:
Yes, I think a metrics util should be in place.